project-code/app/com/edulify/modules/geolocation/GeolocationService.java

```diff
@@ -1,14 +1,27 @@
 package com.edulify.modules.geolocation;
 
-/**
- * @deprecated Use AsyncGeolocationService instead
- */
-@Deprecated
-public class GeolocationService {
+import play.libs.F.Promise;
 
-  private static final long timeout = Config.getMillisecondsOr("geolocation.timeout", 5000);
+import static play.libs.F.Promise.pure;
 
-  public static Geolocation getGeolocation(String ip) {
-    return AsyncGeolocationService.getGeolocation(ip).get(timeout);
+public final class GeolocationService {
+
+  private final GeolocationCache cache;
+  private final GeolocationProvider provider;
+
+  public GeolocationService(GeolocationCache cache, GeolocationProvider provider) {
+    this.cache = cache;
+    this.provider = provider;
+  }
+
+  public Promise<Geolocation> getGeolocation(String ip) {
+    if (null == ip) return pure(null);
+    if (ip.isEmpty()) return pure(null);
+    Geolocation geolocation = cache.get(ip);
```

Comment:
Unfortunately can't see a way to do it because of:

 - Play's Cache stores retrieved object at cache without asking do we need it - too bad
 - Even if code will look some like this
```java
//.... class GeolocationCache ....
  public Geolocation getOrElse(String ip, Callable<Geolocation> block) throws Exception {
    return cacheOn ? Cache.getOrElse(key(ip), block, 0) : block.call();
  }
```
we should use Callable that returns `Geolocation` what means that we can't use `GeolocationProvider` instances because those instances returns `Promise<Geolocation>` and we can't get plain `Geolocation` object (without blocking, that is definitely no that we want)

So I'd prefer to leave it as is. But may be You have a vision/solution for this case?